### PR TITLE
Add user promo codes for submitted twitter handles

### PR DESCRIPTION
**Pull request by IvanIvanoff**
## Changes

Add `user_promo_codes` table that will store the availalbe promo codes for each user.
When a submitted Twitter handle is approved (at a later stage by an admin/moderator), if the total number of submitted handles:
- exceeds 3 - Add a promo code for 27% discount;
- exceeds 6 - Add a promo code for 54% discount.

The user promo codes can be obtained by the `currentUser` API.
```graphql
{
  currentUser {
    id
    promoCodes{
      campaign
      coupon
      percentOff
      timesRedeemed
      maxRedemptions
      
    }
  }
}
```
```json
{
  "data": {
    "currentUser": {
      "id": "3",
      "promoCodes": [
        {
          "campaign": "local_test",
          "coupon": "wo4V68Vu",
          "maxRedemptions": 1,
          "percentOff": 3,
          "timesRedeemed": 0
        }
      ]
    }
  }
}
```

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have tried to find clearer solution before commenting hard-to-understand parts of code
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: lib/sanbase_web/graphql/resolvers/billing_resolver.ex [hunk 1]
```diff
@@ -19,6 +22,11 @@ defmodule SanbaseWeb.Graphql.Resolvers.BillingResolver do
 
     with {:plan?, %Plan{is_deprecated: false} = plan} <- {:plan?, Plan.by_id(plan_id)},
          {:ok, subscription} <- route_subscription(current_user, plan, payment_instrument, coupon) do
+      # If the coupon exists in the user_promo_codes table, times_redeemed
+      # will be bumped by one. We don't check beforehand if the coupon exists and if it's valid,
+      # as the Stripe API will take care of this.
+      if coupon, do: UserPromoCode.use_coupon(coupon)
```
Comment (tspenov) [inline lib/sanbase_web/graphql/resolvers/billing_resolver.ex:28]: we can update with `times_redeemed` from the coupon object itself so it is synced with Stripe ?

**Review by tspenov — APPROVED** [review]
LGTM, small comments

File: lib/sanbase/monitored_twitter_handle/monitored_twitter_handle.ex [hunk 2]
```diff
@@ -71,4 +98,51 @@ defmodule Sanbase.MonitoredTwitterHandle do
...
+
+  defp create_user_promo_code(user_id, percent_off) do
+    redeem_by = DateTime.utc_now() |> DateTime.add(30, :day) |> DateTime.truncate(:second)
+
+    {:ok, coupon} =
+      Sanbase.StripeApi.create_promo_coupon(%{
+        duration: "once",
+        percent_off: percent_off,
+        redeem_by: redeem_by |> DateTime.to_unix(),
+        metadata: %{campaign: @campaign, products: "SANBASE"},
```
Comment (tspenov) [inline lib/sanbase/monitored_twitter_handle/monitored_twitter_handle.ex:133]: Maybe should be `product` as in the local table ?

File: lib/sanbase/billing/user_promo_codes.ex [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+      nil ->
+        # If there is no information in this table, this means that the coupon
+        # is issued in another way, e.g. by Stripe directly. In this case, we
+        # cannot check if the coupon is usable, so we assume it is. If it is not
+        # the step that submits it to the Stripe API will fail
+        true
+
+      %__MODULE__{} = promo ->
+        cond do
+          get_in(promo, [:metadata, "product"]) != plan.product.code ->
```
Comment (tspenov) [inline lib/sanbase/billing/user_promo_codes.ex:88]: maybe check `!=` only if there is something in  `get_in(promo, [:metadata, "product"])`

File: lib/sanbase/monitored_twitter_handle/monitored_twitter_handle.ex [hunk 4]
```diff
@@ -71,4 +98,50 @@ defmodule Sanbase.MonitoredTwitterHandle do
...
+      Sanbase.StripeApi.create_promo_coupon(%{
+        duration: "once",
+        percent_off: percent_off,
+        redeem_by: redeem_by |> DateTime.to_unix(),
+        metadata: %{campaign: "locl_test"},
+        max_redemptions: 1
+      })
+
+    UserPromoCodes.create(%{
+      campaign: "local_test",
```
[the remark this answers is not in the document]
Comment (IvanIvanoff) [reply lib/sanbase/monitored_twitter_handle/monitored_twitter_handle.ex]: cleaned

File: lib/sanbase/monitored_twitter_handle/monitored_twitter_handle.ex [hunk 5]
```diff
@@ -71,4 +98,50 @@ defmodule Sanbase.MonitoredTwitterHandle do
...
+
+  defp create_user_promo_code(user_id, percent_off) do
+    redeem_by = DateTime.utc_now() |> DateTime.add(180, :day) |> DateTime.truncate(:second)
+
+    {:ok, coupon} =
+      Sanbase.StripeApi.create_promo_coupon(%{
+        duration: "once",
+        percent_off: percent_off,
+        redeem_by: redeem_by |> DateTime.to_unix(),
+        metadata: %{campaign: "locl_test"},
```
[the remark this answers is not in the document]
Comment (IvanIvanoff) [reply lib/sanbase/monitored_twitter_handle/monitored_twitter_handle.ex]: cleaned

File: lib/sanbase/monitored_twitter_handle/monitored_twitter_handle.ex (hunk 2 above)
Comment (IvanIvanoff) [reply lib/sanbase/monitored_twitter_handle/monitored_twitter_handle.ex:133]: this is a leftover, as first I tried to do `products: ["sanbase"]`, but stripe does not allow lists.

File: lib/sanbase/billing/user_promo_codes.ex (hunk 3 above)
Comment (IvanIvanoff) [reply lib/sanbase/billing/user_promo_codes.ex]: Done.

File: lib/sanbase_web/graphql/resolvers/billing_resolver.ex (hunk 1 above)
Comment (IvanIvanoff) [reply lib/sanbase_web/graphql/resolvers/billing_resolver.ex:29]: I'll replace this with `sync_with_stripe_data`, but in order to be able to deploy now so the FE can work on it, I'll do it in the next PR.

**Pull request merged**